### DISCUSSION TITLE
NumPy issue during install

### DISCUSSION THREAD

**Discussion by philipnenni** [pydata/sparse#560] [Q&A] [answered]
Can I get some help on resolving this issue? thanks in advance I'm a python newbie.

ERROR: Cannot install numpy==1.21.6 and numpy==1.22.3 because these package versions have conflicting dependencies.

      The conflict is caused by:
          The user requested numpy==1.22.3
          The user requested numpy==1.21.6

**Comment by rgommers** [comment]
It would be helpful to say what command you ran to get that error message @philipnenni

**Reply by philipnenni** [reply]
I was trying to install sparse so the command I used was 'pip install sparse'
I also tried in PyCharm and limiting the version of NumPy but neither option worked.

**Reply by rgommers** [reply]
The only requirements in `sparse` are (from https://github.com/pydata/sparse/blob/master/requirements.txt):
```
numpy>=1.17
scipy>=0.19
numba>=0.49
```

So this seems unlikely to be the cause. Can you provide the complete output of `pip install sparse`, and also say how you installed Python and whether or not you are working in a virtualenv?

**Reply by philipnenni** [reply]
I think you are right... I'm still getting used to reading the logs and whatnot. I'm using Python 3.10.7 on windows.
Here is the full output.

Collecting sparse
  Using cached sparse-0.13.0-py2.py3-none-any.whl (77 kB)
Collecting numpy>=1.17
  Using cached numpy-1.23.3-cp310-cp310-win32.whl (12.2 MB)
Collecting scipy>=0.19
  Using cached scipy-1.9.1.tar.gz (42.0 MB)
  Installing build dependencies ... error
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [34 lines of output]
      Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "aarch64" and platform_python_implementation != "PyPy"' don't match your environment
      Ignoring numpy: markers 'python_version == "3.8" and platform_machine == "arm64" and platform_system == "Darwin"' don't match your environment
      Ignoring numpy: markers 'python_version == "3.9" and platform_machine == "arm64" and platform_system == "Darwin"' don't match your environment
      Ignoring numpy: markers 'platform_machine == "loongarch64"' don't match your environment
      Ignoring numpy: markers 'python_version == "3.8" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "aarch64" and platform_machine != "loongarch64" and platform_python_implementation != "PyPy"' don't match your environment
      Ignoring numpy: markers 'python_version == "3.9" and (platform_machine != "arm64" or platform_system != "Darwin") and platform_machine != "loongarch64" and platform_python_implementation != "PyPy"' don't match your environment
      Ignoring numpy: markers 'python_version == "3.11" and platform_python_implementation != "PyPy"' don't match your environment
      Ignoring numpy: markers 'python_version >= "3.12"' don't match your environment
      Ignoring numpy: markers 'python_version >= "3.8" and platform_python_implementation == "PyPy"' don't match your environment
      Collecting meson-python>=0.8.1
        Using cached meson_python-0.9.0-py3-none-any.whl (18 kB)
      Collecting meson==0.62.2
        Using cached meson-0.62.2-py3-none-any.whl (864 kB)
      Collecting Cython<3.0,>=0.29.21
        Using cached Cython-0.29.32-py2.py3-none-any.whl (986 kB)
      Collecting pybind11<2.11.0,>=2.4.3
        Using cached pybind11-2.10.0-py3-none-any.whl (213 kB)
      Collecting pythran<0.12.0,>=0.9.12
        Using cached pythran-0.11.0-py3-none-any.whl (4.2 MB)
      Collecting wheel<0.38.0
        Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
      Collecting numpy==1.22.3
        Using cached numpy-1.22.3-cp310-cp310-win32.whl (12.2 MB)
      ERROR: Cannot install numpy==1.21.6 and numpy==1.22.3 because these package versions have conflicting dependencies.

      The conflict is caused by:
          The user requested numpy==1.22.3
          The user requested numpy==1.21.6

      To fix this you could try to:
      1. loosen the range of package versions you've specified
      2. remove package versions to allow pip attempt to solve the dependency conflict

      ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

**Reply by rgommers** [reply, marked as answer]
`Using cached scipy-1.9.1.tar.gz` indicates that `pip` is trying to build SciPy from source. What is happening here is that there's no SciPy wheel for 32-bit Windows at https://pypi.org/project/scipy/1.9.1/#files. 

You're not going to be able to resolve that easily. If you are new to Python and can start over, the best thing is to get rid of your 32-bit install completely, and install a 64-bit Python. If you don't know which one, I recommend the 64-bit Windows installer of Anaconda at https://www.anaconda.com/products/distribution. It already comes with NumPy, SciPy, Numba etc. pre-installed (maybe even sparse, no sure). Otherwise, if you prefer the python.org installer, also there 64-bit is recommended: https://www.python.org/downloads/release/python-3107/

**Reply by philipnenni** [reply]
Cool! Thanks for the help. Sorry about my lack of experience and understanding of the logs.

**Reply by rgommers** [reply]
No worries at all!

**Reply by hameerabbasi** [reply]
Thanks @rgommers for replying to this much faster than I could. I appreciate it!
